openedx/core/djangoapps/content_libraries/docs/decisions/0002-libraries-lti-provider.rst

```diff
@@ -0,0 +1,35 @@
+Allow content libraries to be used by LTI consumers
+---------------------------------------------------
+
+Status
+------
+
+Pending
+
+Context
+-------
+
+Currently, there's no way for xblocks in blockstore-based content libraries to be served to other LTI consumers. This ADR explores ways in which this can be achieved, without overly complicating the codebase. The platform currently does have an :code:`lms.lti_provider` app which implements the LTI 1.1 spec, but it only works for modulestore-based courses, and the LTI 1.1 spec itself is also deprecated and deemed insecure.
+
+Decision
+--------
+
+We will use the latest LTI spec (`LTI 1.3`_) to serve the xblocks. This would be largely independent of the existing LTI 1.1 implementation, and would barely require any changes in that app.
+
+Instead, we would add an :code:`lti` package to the :code:`openedx.core.content_libraries` app, which will handle LTI endpoints, and communicate with the LTI consumers. We will use the actively-maintained `pylti1.3`_ package which can handle the LTI-spec implementation, which saves a lot of development and testing effort, and keeps the platform codebase relevant to openedx. The library is maintained by `@dmitry-viskov`_, who also proposed using it in an `earlier PR`_ to add LTI 1.3 support for courses. `pylti1.3`_ exists as an up-to-date `PyPi package`_, and is also used by projects like `my-learning-analytics`_, and `NimbleWise (source)`_.
```

Comment:
What about all the concerns listed in that earlier PR? https://github.com/edx/edx-platform/pull/21435#pullrequestreview-455386016 I think this ADR should address them if we're proposing a similar approach.

Reply:
> What about all the concerns listed in that earlier PR?

Here's a summary of the concerns captured in the aforementioned PR.

1. Reliability and maintainability of the LTI implementation being adopted.
2. Caveats around auto user creation: (a.) User not enrolled in any course, (b.) Linking with existing user accounts, (c.) Authenticated users accessing content from different views.
3. Isolated content during launches.

The ADR already addresss some of the above:

- (1.) The doc is advocating `pylti1p3` as enough by indirectly trusting its usage and credibility in other tools, some of them certified, according to the library author.
- (3.) This is somewhat irrelevant since this ADR is proposing exactly that: Using content libraries to offer and launch isolated content that by definition was designed to be embeddable in isolation of courseware and can be easily rendered inside iframes.

What the ADR doesn't specify is how authorization and authentication will be done for content libraries (topic (2.)).

Let's discuss them on top of the [PR being proposed](https://github.com/edx/edx-platform/pull/27411), which basically follows the same approach of the LTI 1.1.

a. Course enrollement is irrelevant since content is independent.
b. There is no use case for user association, AFAIK. But the current spec doesn't block that either. As mentioned, LTI has some tooling to aid this type of association. Augmenting the implementation to add support for those is possible.
c. There is currently no objection to keeping the same approach used by LTI 1.1 for LTI 1.3. AFAIK, once authenticated, the user could access other "parts" of edx with no impact on functionality.

> I think this ADR should address them if we're proposing a similar approach.

I think it is valuable to include the following:

1. Mention why we are not going to onboard user profile and mapping betwen tool and platform, but acknowledging its existence.
1. Mention the implications of user creation at launch, more likely in the Consequences section.

Does it makes sense to you?

Reply:
Sounds good, thanks.